src/Components/Components/src/Lifetime/ComponentApplicationState.cs

```diff
@@ -0,0 +1,152 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// The state for the component and services of a components application.
+    /// </summary>
+    public class ComponentApplicationState
+    {
+        private IDictionary<string, byte[]>? _existingState;
+        private readonly IDictionary<string, byte[]> _currentState;
+        private readonly List<Func<Task>> _registeredCallbacks;
+
+        internal ComponentApplicationState(
+            IDictionary<string, byte[]> currentState,
+            List<Func<Task>> pauseCallbacks)
+        {
+            _currentState = currentState;
+            _registeredCallbacks = pauseCallbacks;
+        }
+
+        internal void InitializeExistingState(IDictionary<string, byte[]> existingState)
+        {
+            if (_existingState != null)
+            {
+                throw new InvalidOperationException("ComponentApplicationState already initialized.");
+            }
+            _existingState = existingState ?? throw new ArgumentNullException(nameof(existingState));
+        }
+
+        /// <summary>
+        /// Registers a callback that will be invoked when the component application is being paused.
+        /// </summary>
+        /// <param name="callback">The <see cref="Func{TResult}"/> to invoke.</param>
+        public void RegisterOnPersistingCallback(Func<Task> callback)
+        {
+            if (callback is null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            _registeredCallbacks.Add(callback);
+        }
+
+        /// <summary>
+        /// Tries to retrieve the persisted state with the given <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">The key used to persist the state.</param>
+        /// <param name="value">The persisted state.</param>
+        /// <returns><c>true</c> if the state was found;<c>false</c> otherwise.</returns>
+        public bool TryRetrievePersistedState(string key, [MaybeNullWhen(false)] out byte[] value)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (_existingState == null)
+            {
+                // Services during prerendering might try to access their state upon injection on the page
+                // and we don't want to fail in that case.
+                // When a service is prerendering there is no state to restore and in other cases the host
+                // is responsible for initializing the state before services or components can access it.
+                value = null;
+                return false;
+            }
+
+            if (_existingState.TryGetValue(key, out value))
+            {
+                _existingState.Remove(key);
```

Comment:
I'm concerned "auto-remove state on retrieval" is a very strange pattern that will confuse people and make things harder for them. At the very least if we retain this behavior we should change the API name to something like `TryRetrieveAndRemovePersistedState`.

The sort of case where I think this risks making things difficult is if there are multiple instances of a component that make use of the same state. This auto-removal behavior means the developer is forced to invent a system for giving each of them unique keys that they can regenerate during rehydration.

Is there a strict reason why we have to do this? What goes wrong if we leave the state in place?

Reply:
I thought about this and the answer that I have is that multiple instances of a component should not use the same state. The state is private, its for you and you should not "share" it with anyone through the cache.

If you want to share a piece of data like some values you retrieved from the database, the logical option is that a service or top level component participates in the restore process and is the one responsible for restoring the data and sharing it with the rest of the components.

There a couple of drawbacks around not doing this:
* Memory is never released since it's hold forever in the cache, which is not a feature we want to have.
* The behavior is not correct when you do something like a client navigation away and back from the page, since the state is still present there.

Reply:
OK, once I saw the samples, it was a bit clearer to me why the pattern is useful:

```cs
if (!TryRetrieveAndRemoveState(key, out value))
{
    value = await GetStateFromDb();
}
```

I do think it's important that we have the word `Remove` in the API name to make clear this is the expected pattern, otherwise people are not going to find this clear. It would be like if `Dictionary` auto-removed things when you called `Get`.

As for the drawback I was concerned about with sharing states across components, I do still think some people will run into issues with that, but given you don't want the same instance of a component to reuse the stored state more than once (e.g., after client-side navigation) the only alternative I can think of is having some notion of detecting quiescence at which point we discard the state store and only use freshly-loaded data from then on. I know you had a concern about how meaningful it would be to detect quiescence, but we do it for prerendering. Is there a definite reason you think we shouldn't use the same rule to detect it after entering interactive rendering?

Reply:
I've just realised a weird hacky thing that developers are going to do to work around the restriction we're imposing here:

```
if (componentState.TryRetrieveAndRemoveState(key, out value))
{
    // Put the state back so other components can read it too
    componentState.PersistAsJson(key, value);
}
else
{
    value = await GetStateFromDb();
}
```

I don't mean to say this PR should be blocked on it, but we should decide if this is something we want to do anything about. We don't want to be imposing restrictions that have easy workarounds because then not only are our goals not being met, we're creating battles with our customers.

It's also a bit unclear, given you can do this, when exactly you are supposed to persist state. Should you do it after each time state is retrieved from a DB, just in case the application dies unexpectedly? I know that's not our goal, but developers might think that's how it works.

We could either have the `ComponentApplicationState` service have different modes it goes through (e.g., it only switches into "accepting new state" mode once the system begins persisting state), or we could pass two different services into the component: `IComponentStateProvider` and `IComponentStatePersister` which have completely separate storage - one is read-only, the other is write-only.

Reply:
They can't do that, the restored state is kept separate from the current state, so even if you persist it, other components won't be able to see it.

Reply:
I found a cheap way of also "Clearing" the state after the first render, so I've done that since it's better for SSB and also on Wasm for consistency. That said, I'm keeping the current behavior because I don't want to promote "sharing" of state across components.

The current approach is more conservative and enables us to relax this later on if we want without it being a breaking change.